staging/src/k8s.io/apiserver/pkg/apis/apiserverinternal/types.go

```diff
@@ -0,0 +1,114 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserverinternal
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+//  Storage version of a specific resource.
+type StorageVersion struct {
+	metav1.TypeMeta
+	// The name is <group>.<resource>.
+	metav1.ObjectMeta
+
+	// Spec is an empty spec. It is here to comply with Kubernetes API style.
+	Spec StorageVersionSpec
+
+	// API server instances report the version they can decode and the version they
+	// encode objects to when persisting objects in the backend.
+	Status StorageVersionStatus
+}
+
+// StorageVersionSpec is an empty spec.
+type StorageVersionSpec struct{}
+
+// API server instances report the versions they can decode and the version they
+// encode objects to when persisting objects in the backend.
+type StorageVersionStatus struct {
+	// The reported versions per API server instance.
+	// +optional
+	ServerStorageVersions []ServerStorageVersion
+	// If all API server instances agree on the same encoding storage version,
+	// then this field is set to that version. Otherwise this field is left empty.
+	// +optional
+	AgreedEncodingVersion *string
+
+	// The latest available observations of the storageVersion's state.
+	// +optional
+	Conditions []StorageVersionCondition
+}
+
+// An API server instance reports the version it can decode and the version it
+// encodes objects to when persisting objects in the backend.
+type ServerStorageVersion struct {
+	// The ID of the reporting API server.
+	// For a kube-apiserver, the ID is configured via a flag.
+	APIServerID string
+
+	// The API server encodes the object to this version when persisting it in
+	// the backend (e.g., etcd).
+	EncodingVersion string
+
+	// The API server can decode objects encoded in these versions.
+	// The encodingVersion must be included in the decodableVersions.
+	DecodableVersions []string
+}
+
+type StorageVersionConditionType string
+
+const (
+	// Indicates that encoding storage versions reported by all servers are equal.
+	AllEncondingVersionsEqual StorageVersionConditionType = "AllEncodingVersionsEqual"
+)
+
+type ConditionStatus string
+
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
+// Describes the state of the storageVersion at a certain point.
+type StorageVersionCondition struct {
```

Comment:
match the definition in https://github.com/kubernetes/enhancements/pull/1624 if you would.